### DISCUSSION TITLE
windowing functions could return NULL for partial windows

### DISCUSSION THREAD

**Discussion by jangorecki** [duckdb/duckdb#8340] [General]
This seems to be the default behavior in various software I have used for calculation of rolling statistics.
I think there should be at least an option to enable that behavior in duckdb. As of now partial window is computed and numbers are being returned for an incomplete window, which is rather error-prone.

Taking example from documentation: https://duckdb.org/docs/sql/window_functions.html
```sql
SELECT points,
    SUM(points) OVER (
        ROWS BETWEEN 1 PRECEDING
                 AND 1 FOLLOWING) we
FROM results
```

gives

![moving-sum](https://github.com/duckdb/duckdb/assets/3627377/9faa76df-16b1-4487-a575-7dd8ce4e6428)

In the left most table I would expect the following values

points | we
--- | ---
10 | NULL
8 | 30
12 | 29
... | ...

NULL instead of '18' because window was still partial at that row location (2 out of 3 observations were available).

**Comment by Tishj** [comment]
It took me a little bit to follow, but you're referring to this section right?

> Notice that at the edge of the partition, there are only two values added together. This is because frames are cropped to the edge of the partition.

You basically want the option to make the frame exact, for lack of a better word.

If the frame is cropped (at the edges of the partition) the result should be NULL

**Reply by jangorecki** [reply]
Yes, for me (and many other software) is very  natural.

If I ask for SMA 30 days I want either SMA 30 days or NULL, not an SMA 29 days or SMA 2 days.

**Reply by Tishj** [reply]
I see, maybe a temporary workaround would be to do a count followed by a case to replace the result with null if the frame was only partially filled

```sql
SELECT points,
    count(points) over (
        ROWS BETWEEN 1 PRECEDING
                AND 1 FOLLOWING) as "frame_size",
    frame_size = 3 as "partial",

    case when partial = true then SUM(points) OVER (
        ROWS BETWEEN 1 PRECEDING
                 AND 1 FOLLOWING)
    else null end we
FROM results;
```

This worked, though I couldn't get it to work as a subquery for some reason
```sql
SELECT
    points,
    (
        select we from (
            select count(points) over (
                ROWS BETWEEN 1 PRECEDING
                        AND 1 FOLLOWING) as "frame_size",
            frame_size = 3 as "partial",

            case when partial = true then SUM(points) OVER (
                ROWS BETWEEN 1 PRECEDING
                        AND 1 FOLLOWING)
            else null end we
        FROM results
    )) as we
FROM results;
```

But either way this would only serve as a temporary work around, I think this is a fair point

**Reply by jangorecki** [reply]
it is some workaround, but as I am benchmarking rolling functions I would like to have a solution that does not add extra overhead.

**Reply by jangorecki** [reply]
So is there a way to turn this discussion into an issue, or other kind of ticket that can be more easily considered as one required an action, and more easily tracked?

**Comment by jangorecki** [comment]
I am preparing slides for a talk about windowing functions and would like to know if feature requested here is considered in-scope or out-of-scope for duckdb?

@hawkfish any idea?

if it's in-scope, then could we get an issue created for tracking status of this coming feature, so I can easily link it?

**Reply by hawkfish** [reply]
I don't think it is out of scope exactly, but it is not as simple as described here. This is a special case of fixed (constant) row framing under total ordering and would require a new modifier of some kind. That modifier would then have to interact with a lot of other functionality (exclude, distinct, range frames, partially ordered frames...). It would help if you knew of a SQL dialect/database that supported this functionality?

And for the overhead concern, counting frame sizes within a shared partitioning is extremely fast compared to everything else that is going on per row (e.g., hashing, sorting, segment tree construction, ...) as it just subtracts the frame offsets and stores them in a pre-allocated buffer. I'd give it a try and see what the impact is. At the very least, I've been thinking about sharing frame computations between functions, which would make this even faster, and that is a lot more likely to happen in the short term.

**Reply by jangorecki** [reply]
I don't recall any SQL dialect. I elaborated a bit more at https://github.com/PRQL/prql/issues/3618, you may find it useful
> My observation is that SQL databases are generally not giving any flexibility in that matter, and always compute and return partial window. On the other hand analytical tools are generally not only giving such flexibility but most of them defaults to not returning partial window results. What behavior is desires depends on the use case. From my personal uses cases, it is definitely the latter one that was useful.

so, can we have an issue opened to track this FR?

**Reply by hawkfish** [reply]
The pandas `min_periods` is roughly what I meant by a new keyword, and the observation in that thread that SQL defaults to 1 is helpful for defining semantics for feature interactions.

Feel free to turn this into an issue, but be aware that we have a bot that clears out issues with 90 days of inactivity, and I don't think anyone is likely to look at it in that time frame.

**Reply by jangorecki** [reply]
uh OK, then sounds it doesn't make sense to keep bumping an issue every 90 days to keep it open...

**Comment by hawkfish** [comment]
Finally got around to benchmarking suggestions:

```sql
SELECT SETSEED(0.8675309);
CREATE OR REPLACE TABLE results AS 
	SELECT ROUND(random() * 20, 0)::TINYINT AS points
	FROM range(10_000_000);

SELECT points,
    SUM(points) OVER w AS we,
FROM results
WINDOW w AS (ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING)
```
| run | timing |
| --- | --- |
| 1 | 0.989795 |
| 2 | 0.980213 |
| 3 | 0.910735 |
| 4 | 0.911356 |
| 5 | 0.961919 |

```sql
SELECT points,
    SUM(points) OVER w AS we,
    COUNT(*) OVER w as n,
FROM results
WINDOW w AS (ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING)
```

| run | timing |
| --- | --- |
| 1 | 1.203144 |
| 2 | 1.191426 |
| 3 | 1.102619 |
| 4 | 1.085473 |
| 5 | 1.156939 |

```sql
WITH variable AS (
	SELECT points,
		SUM(points) OVER w AS we,
		COUNT(*) OVER w as n,
	FROM results
	WINDOW w AS (ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING)
)
SELECT points,
	IF(n < 3, NULL, we) AS we,
FROM variable
```

| run | timing |
| --- | --- |
| 1 | 1.168134 |
| 2 | 1.112830 |
| 3 | 1.176201 |
| 4 | 1.103230 |
| 5 | 1.118889 |

So adding the count adds 20% overhead but using it to recompute adds nothing.

**Reply by jangorecki** [reply]
Thank you for benchmarking, it would be nice to know if numbers are seconds, or milliseconds, etc. Also 2-3 different data sizes could present how it is scaling with size increase.

**Reply by hawkfish** [reply]
Sorry, seconds (all our benchmark runner results are in seconds).

I used `TINYINT` because that was the size of the data in your example, but feel free to [try your own benchmarks](https://duckdb.org/docs/dev/benchmark).

One other thing that occurred to me this morning: This proposal is not well-defined for `RANGE` framing because you don't know what the "missing" boundary values are (they could be inside or outside the frame). So I suspect that the reason the standard does not have a way to specify this is because it does not apply universally, and the count workaround is easy to implement. 

It might be worth looking into special-casing `COUNT(*)` with no `EXCLUDE` clause - the current implementation calls a function for each row, but a vectorised loop would likely be much faster.

**Comment by hawkfish** [comment]
Tried out vectorising `COUNT(*)` and the time was halfway between the old `COUNT(*)` and no `COUNT(*)`. So an improvement, but still about 10% slower.

**Reply by hawkfish** [reply]
The other thing that might help here is sharing frame boundaries. I have plans to add that, but it's not the sort of thing I can check quickly.
